packages/runtime/container-runtime/src/opLifecycle/batchManager.ts

```diff
@@ -169,14 +174,20 @@ export class BatchManager {
 		const startPoint = this.pendingBatch.length;
 		return {
 			rollback: (process: (message: BatchMessage) => void) => {
-				for (let i = this.pendingBatch.length; i > startPoint; ) {
-					i--;
-					const message = this.pendingBatch[i];
+				const rollbackOps = this.pendingBatch.splice(startPoint).reverse();
+				for (const message of rollbackOps) {
 					this.batchContentSize -= message.contents?.length ?? 0;
 					process(message);
 				}
-
-				this.pendingBatch.length = startPoint;
+				const count = this.pendingBatch.length - startPoint;
+				if (count !== 0) {
```

Comment:
Nice, I noticed this gap too.  I was thinking about another case that would hit this, where the batch could be popped (emptied) during the checkpoint, and if we threw afterwards then it gets messed up.  Not that I think this would happen.